subgraphs/_reference_/src/common/utils/logger.ts

```diff
@@ -0,0 +1,26 @@
+import { ethereum, log } from "@graphprotocol/graph-ts";
+
+export class Logger {
+  event: ethereum.Event;
+
+  constructor(event: ethereum.Event) {
+    this.event = event;
+  }
+
+  private formatLog(funcName: string, msg: string): string {
+    return `messari_log: [${funcName}}] tx: ${this.event.transaction.hash.toHexString()} :: ${msg}`;
+  }
+
+  info(funcName: string, msg: string): void {
```

Comment:
I'd keep the ability of passing variables from the user to be formatted too. Otherwise they need to take care of the formatting to string before logging.

```
info(funcName: string, msg: string, params: Array<string>): void {
  log.info(this.formatLog(funcName, msg), params);
}
```

Also, since we are already used to the format `log.info(template, params)`, I'd make the funcName be the last parameter (so people don't get confused and accidentally put the message there).

Or! Even better. When you use the logger you have to instantiate it first, right? Make the function where it is instantiated one of the constructor arguments. Then we don't need to set the name every time we log. The only issue is if you pass the logger down to other aux functions in you handler. I can think of a hack for this, but feels a bit ugly, like fighting against the dev 😅 . But here it goes, for inspiration:

```
class Logger { // unexported, so we don't have access to the type, this way we cannot make it a function argument
  wrap(): RefLogger {}
}

export MakeLogger(...); // the constructor to be able to instantiate the logger in the first place

export class RefLogger {
  private logger: Logger
  getLogger(fnName: string): Logger;
}
```

This way, if we want to use it in mutliple functions we'd need to do something like:
```
export function handleEvent(ev: Event): void {
  const logger = MakeLogger("handleEvent", ev, ....);
  logger.info("something something");
  someAuxStuff(ev, logger.wrap());
}

function someAuxStuff(ev: Event, log: RefLogger): void { // RefLogger because since Logger is not exported we can only pass it this way
  const logger = log.getLogger("someAuxStuff");
  
  logger.warning("uh, bad things happening");
  moreAuxStuff(ev, logger.wrap());
}

function moreAuxStuff(ev: Event, log: RefLogger): void {
  const logger = log.getLogger("moreAuxStuff");
  // ...
}
```

It could happen that you pass the logger down a few levels but only use it in the inner most call. Since you need to unwrap it with a function name, we might not have the whole stack, but at least the function name where the message was logged.

Reply:
lmk what you think of the changes I made

Reply:
I like it